Tests/Test.cs

```diff
@@ -0,0 +1,327 @@
+        repository.GetAndStore(
+            id,
+            model => decider.Decide(command, Map(model)),
+            ct
+        );
+
+    private ShoppingCart Map(ShoppingCartModel? model)
+    {
+        if (model == null)
+            return new Empty();
+
+        switch (model.Status)
+        {
+            case ShoppingCartModel.ShoppingCartStatus.Pending:
+                return new Pending(
+                    model.ProductItems.ToImmutableDictionary(ks => $"{ks.ProductId}{ks.Price}", vs => vs.Quantity)
+                );
+
+            case ShoppingCartModel.ShoppingCartStatus.Confirmed:
+                return new Closed();
+
+            default:
+                throw new ArgumentOutOfRangeException();
+        }
+    }
+}
+
+public class ECommerceDBContext: DbContext
+{
+    public DbSet<ShoppingCartModel> ShoppingCarts { get; set; } = default!;
+
+    protected override void OnModelCreating(ModelBuilder modelBuilder)
+    {
+        modelBuilder.Entity<ShoppingCartModel>()
+            .ToTable("ShoppingCarts")
+            .OwnsMany(b => b.ProductItems, a =>
+            {
+                a.ToTable("ProductItems");
+                a.WithOwner().HasForeignKey("ShoppingCartId");
+                a.HasKey("ProductId", "ShoppingCartId");
+            });
+    }
+
+    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder) =>
+        optionsBuilder
+            .UseInMemoryDatabase("ECommerceTest");
+}
+
+public class ShoppingCartRepository(ECommerceDBContext dbContext): IShoppingCartRepository
+{
+    public ValueTask<ShoppingCartModel?> Find(Guid id, CancellationToken ct) =>
+        dbContext.FindAsync<ShoppingCartModel>([id], ct);
+
+    public Task Store(Guid id, ShoppingCartModel? current, ShoppingCartEvent[] events, CancellationToken ct)
+    {
+        foreach (var @event in events)
+        {
+            switch (@event, current)
+            {
+                case (Opened(var clientId, var openedAt), null):
+                    dbContext.Add(new ShoppingCartModel
+                    {
+                        Id = id,
+                        ClientId = clientId,
+                        Status = ShoppingCartModel.ShoppingCartStatus.Pending,
+                        OpenedAt = openedAt
+                    });
+                    break;
+                case (ProductAdded(var (productId, quantity, price), var addedAt), not null):
+                    var toUpdate = current.ProductItems.SingleOrDefault(p => p.ProductId == productId);
+
+                    if (toUpdate == null)
+                    {
+                        current.ProductItems.Add(new ShoppingCartModel.PricedProductItem
+                        {
+                            ProductId = productId, Quantity = quantity, Price = price, AddedAt = addedAt,
+                        });
+                    }
+                    else
+                    {
+                        toUpdate.Quantity += quantity;
```

Comment:
what if the price changed (I think the price thing is just broken but...)
also is it price or total or what

Reply:
The price won't change here as it's frozen in the event.